src/get-started/write-and-run-unit-tests.md

```diff
@@ -1,10 +1,229 @@
 # Write and run unit tests
 
-including but not limited to:
+The TiDB project runs unit tests using [testify](https://github.com/stretchr/testify).
 
-* tidb unit test framework
-* how to write unit tests
-* how to run all the tests
-* how to run a single test
-* how to find the failed test
+You may find tests using [pingcap/check](http://github.com/pingcap/check) which is a fork of [go-check/check](https://github.com/go-check/check), but since that framework is poorly maintained, we are migrating to testify.
 
+You can check the background and progress on the migration [tracking issue](https://github.com/pingcap/tidb/issues/26022).
+
+## How to write unit tests?
+
+We uses testify for writing unit tests. Basically, it is out-of-the-box [testing](https://pkg.go.dev/testing) with testify assertions.
+
+### TestMain
+
+When running tests, Golang compiles each package along with any files with names with suffix `_test.go`. Thus, a test binary contains tests in a package.
+
+Golang testing provides a mechanism to support doing extra setup or teardown before or after testing by writing a package level unique function:
+
+```go
+func TestMain(m *testing.M)
+```
+
+We leverage the function to detect Goroutine leak by [goleak](https://github.com/uber-go/goleak) after all tests finished.
+
+Before writing any unit tests in a package, create a file named `main_test.go` and setup the scaffolding:
+
+```go
+func TestMain(m *testing.M) {
+    goleak.VerifyTestMain(m)
+}
+```
+
+You can also put global variables or helper functions of the test binary in this file.
+
+### Assertion
+
+Let's write a basic test for utility function `StrLenOfUint64Fast`.
+
+```go
+func TestStrLenOfUint64Fast(t *testing.T) {
+    for i := 0; i < 1000000; i++ {
+        num := rand.Uint64()
+        expected := len(strconv.FormatUint(num, 10))
+        actual := StrLenOfUint64Fast(num)
+        require.Equal(t, expected, actual)
+    }
+}
+```
+
+Golang testing detects test functions from `*_test.go` files of the form:
+
+```go
+func TestXxx(*testing.T)
+```
+
+where `Xxx` does not start with a lowercase letter. The function name serves to identify the test routine.
+
+We follow this pattern but using testify assertions instead of out-of-the-box methods, like `Error` or `Fail`, since they are too low level to use.
+
+Almost we use `require.Xxx` for assertions, which imported from `github.com/stretchr/testify/require`. It fails immediately if the assertions fails and we tend to fail tests fast.
+
+Below are the most frequently used assertions:
+
+```go
+func Equal(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{})
+func EqualValues(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{})
+func Len(t TestingT, object interface{}, length int, msgAndArgs ...interface{})
+func NoError(t TestingT, err error, msgAndArgs ...interface{})
+func True(t TestingT, value bool, msgAndArgs ...interface{})
```

Comment:
`func NotNil(t TestingT, object interface{}, msgAndArgs ...interface{}) `
and 
`func Nil(t TestingT, object interface{}, msgAndArgs ...interface{})`
are also frequently used assertions.

Reply:
Updated. Though I think it is enough with these two. Here I aim to show some examples, instead of a full manual of assertion functions.